adafruit_requests.py

```diff
@@ -49,263 +50,537 @@
+                chunk = bytes(b)
+            yield chunk
+        self.close()
 
 
-# pylint: disable=too-many-branches, too-many-statements, unused-argument, too-many-arguments, too-many-locals
-def request(method, url, data=None, json=None, headers=None, stream=False, timeout=1):
-    """Perform an HTTP request to the given url which we will parse to determine
-    whether to use SSL ('https://') or not. We can also send some provided 'data'
-    or a json dictionary which we will stringify. 'headers' is optional HTTP headers
-    sent along. 'stream' will determine if we buffer everything, or whether to only
-    read only when requested
-    """
-    global _the_interface  # pylint: disable=global-statement, invalid-name
-    global _the_sock  # pylint: disable=global-statement, invalid-name
-
-    if not headers:
-        headers = {}
-
-    try:
-        proto, dummy, host, path = url.split("/", 3)
-        # replace spaces in path
-        path = path.replace(" ", "%20")
-    except ValueError:
-        proto, dummy, host = url.split("/", 2)
-        path = ""
-    if proto == "http:":
-        port = 80
-    elif proto == "https:":
-        port = 443
-    else:
-        raise ValueError("Unsupported protocol: " + proto)
-
-    if ":" in host:
-        host, port = host.split(":", 1)
-        port = int(port)
-
-    addr_info = _the_sock.getaddrinfo(host, port, 0, _the_sock.SOCK_STREAM)[0]
-    sock = _the_sock.socket(addr_info[0], addr_info[1], addr_info[2])
-    resp = Response(sock)  # our response
-
-    sock.settimeout(timeout)  # socket read timeout
-
-    try:
+class Session:
+    """HTTP session that shares sockets and ssl context."""
+
+    def __init__(self, socket_pool, ssl_context=None):
+        self._socket_pool = socket_pool
+        self._ssl_context = ssl_context
+        # Hang onto open sockets so that we can reuse them.
+        self._open_sockets = {}
+        self._socket_free = {}
+        self._last_response = None
+
+    def _free_socket(self, socket):
+
+        if socket not in self._open_sockets.values():
+            raise RuntimeError("Socket not from session")
+        self._socket_free[socket] = True
+
+    def _free_sockets(self):
+        free_sockets = []
+        for sock in self._socket_free:
+            if self._socket_free[sock]:
+                sock.close()
+                free_sockets.append(sock)
+        for sock in free_sockets:
+            del self._socket_free[sock]
+            key = None
+            for k in self._open_sockets:
+                if self._open_sockets[k] == sock:
+                    key = k
+                    break
+            if key:
+                del self._open_sockets[key]
+
+    def _get_socket(self, host, port, proto, *, timeout=1):
+        key = (host, port, proto)
+        if key in self._open_sockets:
+            sock = self._open_sockets[key]
+            if self._socket_free[sock]:
+                self._socket_free[sock] = False
```

Comment:
Not sure if I'm misreading - aren't all opened sockets in-use already marked in `_socket_free` as `False` on L416?

Reply:
L416 covers the case where we make a new socket. This spot handles reused sockets that were marked free on line 361.